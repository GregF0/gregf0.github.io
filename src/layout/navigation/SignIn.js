import React, { useState } from 'react'
import { useDispatch } from 'react-redux'
import styled from 'styled-components'
import Button from '@material-ui/core/Button'
import Menu from '@material-ui/core/Menu'
import MenuItem from '@material-ui/core/MenuItem'
import FaceIcon from '@material-ui/icons/Face'
import ListItemIcon from '@material-ui/core/ListItemIcon'
import ListItemText from '@material-ui/core/ListItemText'
import SvgIcon from '@material-ui/core/SvgIcon'
import { GoogleIcon } from './GoogleIcon'
import { linkAccount } from '../../data/auth/actions'
import { googleAuthProvider } from '../../firebase/firebase'
import { setProgressVisibility } from '../../data/progress/actions'

export const SignIn = () => {
  const [anchorEl, setAnchorEl] = useState(null)
  const dispatch = useDispatch()

  const googleAuth = async () => {
    dispatch(setProgressVisibility(true))
    handleClose()
    await dispatch(linkAccount(googleAuthProvider))
    dispatch(setProgressVisibility(false))
  }

  const openMenu = (event) => {
    setAnchorEl(event.currentTarget)
  }

  const handleClose = () => {
    setAnchorEl(null)
  }

  return (
    <>
      <TriggerButton
        aria-controls="sign-in-menu"
        aria-haspopup="true"
        onClick={openMenu}
        startIcon={<FaceIcon />}
      >
        Sign In
      </TriggerButton>
      <Menu
        id="sign-in-menu"
        anchorEl={anchorEl}
        keepMounted
        open={!!anchorEl}
        onClose={handleClose}
      >
        <MenuItem onClick={googleAuth}>
          <ListItemIcon>
            <SvgIcon>
              <GoogleIcon />
            </SvgIcon>
          </ListItemIcon>
          <ListItemText primary="Sign in via Google" />
        </MenuItem>
      </Menu>
    </>
  )
}

export const TriggerButton = styled(Button)`
  color: #fff;
`
