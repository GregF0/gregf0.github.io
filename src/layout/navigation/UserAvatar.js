import React, { useState } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import styled, { css } from 'styled-components'
import useMediaQuery from '@material-ui/core/useMediaQuery'
import Menu from '@material-ui/core/Menu'
import MenuItem from '@material-ui/core/MenuItem'
import MatAvatar from '@material-ui/core/Avatar'
import { startSignOut } from '../../data/auth/actions'
import { setSettings } from '../../data/settings/actions'
import { initialState } from '../../data/settings/reducer'

export const UserAvatar = () => {
  const { name, photo } = useSelector((state) => state.auth)
  const [anchorEl, setAnchorEl] = useState(null)
  const dispatch = useDispatch()

  const signOut = () => {
    dispatch(startSignOut())
    dispatch(setSettings(initialState))
    handleClose()
  }

  const openMenu = (event) => {
    setAnchorEl(event.currentTarget)
  }

  const handleClose = () => {
    setAnchorEl(null)
  }

  const sidenav = +useMediaQuery('(min-width:600px) and (min-height:500px)')

  return (
    <>
      <Avatar sidenav={sidenav} alt={name} src={photo} onClick={openMenu} />
      <Menu
        id="user-menu"
        anchorEl={anchorEl}
        keepMounted
        open={!!anchorEl}
        onClose={handleClose}
      >
        <MenuItem onClick={signOut}>Sign Out</MenuItem>
      </Menu>
    </>
  )
}

export const Avatar = styled(MatAvatar)`
  width: 30px;
  height: 30px;
  cursor: pointer;

  ${({ sidenav }) =>
    sidenav &&
    css`
      width: 35px;
      height: 35px;
    `}
`
