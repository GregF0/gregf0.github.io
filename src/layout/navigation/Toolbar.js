import React, { useEffect, useState } from 'react'
import { useSelector } from 'react-redux'
import styled, { css } from 'styled-components'
import { useMediaQuery } from '@material-ui/core'
import MatToolbar from '@material-ui/core/Toolbar'
import Box from '@material-ui/core/Box'
import { SignIn } from './SignIn'
import { UserAvatar } from './UserAvatar'
import OfflineIcon from '@material-ui/icons/WifiOff'
import logo from './assets/logo.svg'

export const Toolbar = () => {
  const isUserPerm = useSelector((state) => state.auth.name)
  const [isOnline, setIsOnline] = useState(navigator.onLine)
  const mdScreen = +useMediaQuery('(min-width:600px) and (min-height:500px)')

  useEffect(() => {
    const handleOnline = () => {
      setIsOnline(true)
    }

    const handleOffline = () => {
      setIsOnline(false)
    }

    window.addEventListener('online', handleOnline)
    window.addEventListener('offline', handleOffline)

    return () => {
      window.removeEventListener('online', handleOnline)
      window.removeEventListener('offline', handleOffline)
    }
  }, [])

  return (
    <MatToolbar style={{ paddingTop: 'env(safe-area-inset-top)' }}>
      <Box
        width="100%"
        display="flex"
        justifyContent="space-between"
        alignItems="center"
      >
        <Logo mdscreen={mdScreen} src={logo} alt="Pomodor logo" />

        {isOnline ? (
          isUserPerm ? (
            <UserAvatar />
          ) : (
            <SignIn />
          )
        ) : (
          <OfflineIcon data-role="offline-icon" />
        )}
      </Box>
    </MatToolbar>
  )
}

export const Logo = styled.img`
  height: 26px;

  ${({ mdscreen }) =>
    mdscreen &&
    css`
      height: 33px;
    `}
`
