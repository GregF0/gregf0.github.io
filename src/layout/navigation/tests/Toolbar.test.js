import React from 'react'
import * as redux from 'react-redux'
import { createShallow } from '@material-ui/core/test-utils'
import OfflineIcon from '@material-ui/icons/WifiOff'
import { Toolbar, Logo } from '../Toolbar'
import { UserAvatar } from '../UserAvatar'
import { SignIn } from '../SignIn'

describe('<Toolbar />', () => {
  const shallow = createShallow()
  const createWrapper = () => {
    return shallow(<Toolbar />)
  }

  const createStore = (name = null) => {
    const store = {
      auth: { name },
    }

    jest
      .spyOn(redux, 'useSelector')
      .mockImplementation((callback) => callback(store))

    return store
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  test('should render <Toolbar /> correctly when the user is online', () => {
    createStore()
    expect(createWrapper()).toMatchSnapshot()
  })

  test('should render <UserAvatar /> when the user is online and logged in', () => {
    jest.spyOn(navigator, 'onLine', 'get').mockReturnValue(true)
    createStore('Sasha')

    const wrapper = createWrapper()

    expect(wrapper.find(UserAvatar)).toHaveLength(1)
    expect(wrapper.find(OfflineIcon)).toHaveLength(0)
    expect(wrapper.find(SignIn)).toHaveLength(0)
  })

  test('should render <SignIn /> when the user is online and not logged in', () => {
    jest.spyOn(navigator, 'onLine', 'get').mockReturnValue(true)
    createStore()

    const wrapper = createWrapper()

    expect(wrapper.find(SignIn)).toHaveLength(1)
    expect(wrapper.find(OfflineIcon)).toHaveLength(0)
    expect(wrapper.find(UserAvatar)).toHaveLength(0)
  })

  test('should render <Toolbar /> correctly when the user is offline', () => {
    jest.spyOn(navigator, 'onLine', 'get').mockReturnValueOnce(false)
    createStore()

    expect(createWrapper()).toMatchSnapshot()
  })

  test('should render offline icon when the user is offline', () => {
    jest.spyOn(navigator, 'onLine', 'get').mockReturnValue(false)
    createStore()

    const wrapper = createWrapper()

    expect(wrapper.find(OfflineIcon)).toHaveLength(1)
    expect(wrapper.find(UserAvatar)).toHaveLength(0)
    expect(wrapper.find(SignIn)).toHaveLength(0)
  })

  test('should render Pomodor logo', () => {
    createStore()
    expect(createWrapper().find(Logo)).toHaveLength(1)
  })
})
