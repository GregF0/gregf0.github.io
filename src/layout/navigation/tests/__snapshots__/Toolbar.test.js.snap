// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toolbar /> should render <Toolbar /> correctly when the user is offline 1`] = `
<WithStyles(ForwardRef(Toolbar))>
  <Styled(MuiBox)
    alignItems="center"
    display="flex"
    justifyContent="space-between"
    width="100%"
  >
    <styled.img
      alt="Pomodor logo"
      mdscreen={0}
      src="logo.svg"
    />
    <WifiOffIcon
      data-role="offline-icon"
    />
  </Styled(MuiBox)>
</WithStyles(ForwardRef(Toolbar))>
`;

exports[`<Toolbar /> should render <Toolbar /> correctly when the user is online 1`] = `
<WithStyles(ForwardRef(Toolbar))>
  <Styled(MuiBox)
    alignItems="center"
    display="flex"
    justifyContent="space-between"
    width="100%"
  >
    <styled.img
      alt="Pomodor logo"
      mdscreen={0}
      src="logo.svg"
    />
    <SignIn />
  </Styled(MuiBox)>
</WithStyles(ForwardRef(Toolbar))>
`;
